File: 08-安全/02-网络安全/DNS缓存投毒与RFC5452/python/dnspoison.py
"""DNS 缓存投毒（伪造应答）与 RFC 5452 的组合难度模型。

RFC 5452《Measures for Making DNS More Resilient against Forged Answers》
§7.1 定义符号 I / P / N / F / R / W / D，§7.2 给出：

    P_s   = D * F / (N * P * I)        F = R * W
    P_cs  = 1 - (1 - P_s) ** A         A = T / TTL

并声明代入常用值后化简为 P_cs = 1 - (1 - R/1638400) ** (T/TTL)。

本模块把公式、化简常数与文档里给出的每一组数字都做成可实跑的断言；
文档自身存在多处口径不一致（§7 vs §8.1），一律**记录不改正**。
"""

# ---------------------------------------------------------------- 核心公式

def p_s(fake_packets, ns_count=2.5, ports=1, id_space=65536, outstanding=1):
    """单次窗口内被投毒的概率 P_s = D*F/(N*P*I)。"""
    return outstanding * fake_packets / float(ns_count * ports * id_space)


def p_s_from_rate(rate_pps, window_s=0.1, ns_count=2.5, ports=1,
                  id_space=65536, outstanding=1):
    """F = R*W 后的 P_s。"""
    return p_s(rate_pps * window_s, ns_count, ports, id_space, outstanding)


def p_cs(period_s, ttl_s, rate_pps=None, window_s=0.1, ns_count=2.5, ports=1,
         id_space=65536, outstanding=1, p_single=None, attempts=None):
    """组合成功概率 P_cs = 1 - (1 - P_s) ** A，A = T/TTL。

    TTL == 0（或 None）时按 §7.2 的退化口径：有效 TTL 取 W。
    """
    if p_single is None:
        p_single = p_s_from_rate(rate_pps, window_s, ns_count, ports,
                                 id_space, outstanding)
    if attempts is None:
        eff_ttl = ttl_s if ttl_s else window_s
        attempts = period_s / float(eff_ttl)
    return 1.0 - (1.0 - p_single) ** attempts


def reduced_constant(ns_count=2.5, ports=1, id_space=65536, window_s=0.1,
                     outstanding=1):
    """§7.2 化简后的分母：N*P*I/(D*W) —— 文档给的是 1638400。"""
    return ns_count * ports * id_space / (outstanding * window_s)


# ---------------------------------------------------------------- 反解

def period_for_probability(target, ttl_s, rate_pps=None, window_s=0.1,
                           ns_count=2.5, ports=1, id_space=65536,
                           outstanding=1, p_single=None):
    """达到 target 组合概率所需的时间（秒）。

    用精确公式反解：A = ln(1-target)/ln(1-P_s)，T = A * 有效TTL。
    """
    import math
    if p_single is None:
        p_single = p_s_from_rate(rate_pps, window_s, ns_count, ports,
                                 id_space, outstanding)
    eff_ttl = ttl_s if ttl_s else window_s
    attempts = math.log(1.0 - target) / math.log(1.0 - p_single)
    return attempts * eff_ttl


def rate_for_probability(target, period_s, ttl_s, window_s=0.1, ns_count=2.5,
                         ports=1, id_space=65536, outstanding=1):
    """达到 target 组合概率所需的发包速率（pps），精确公式反解。"""
    eff_ttl = ttl_s if ttl_s else window_s
    attempts = period_s / float(eff_ttl)
    p_single = 1.0 - (1.0 - target) ** (1.0 / attempts)
    return p_single * ns_count * ports * id_space / (outstanding * window_s)


# ---------------------------------------------------------------- 带宽换算

DNS_RESPONSE_BYTES = 80


def bandwidth_bps(rate_pps, packet_bytes=DNS_RESPONSE_BYTES):
    return rate_pps * packet_bytes * 8.0


def bps_to_human(v):
    for unit, div in (("Gb/s", 1e9), ("Mbit/s", 1e6), ("kbit/s", 1e3)):
        if v >= div:
            return v / div, unit
    return v, "bit/s"


# ---------------------------------------------------------------- 常用派生量

def mean_tries_single_shot(ns_count=2.5, ports=1, id_space=65536):
    """§7 首段：匹配 ID + 源/目的地址平均需要的包数 = N*P*I/2。"""
    return ns_count * ports * id_space / 2.0


def birthday_p_s(base_p_s, outstanding):
    """§5 生日攻击：D 个相同的在途查询让每个伪造包的命中机会线性提高。

    文档明确「This assumption only holds for small values of 'D'」，
    所以这里就是线性放大，不做成 1-(1-p)^D。
    """
    return base_p_s * outstanding


# ---------------------------------------------------------------- 防御项

def problem_space(ns_count=2.5, ports=1, id_space=65536):
    return ns_count * ports * id_space


def port_randomization_gain(ports=64000):
    """源端口随机化把问题空间放大的倍数。"""
    return ports

File: 08-安全/02-网络安全/DNS缓存投毒与RFC5452/go/dnspoison.go
// Package dnspoison 复现 RFC 5452 §7.2 的组合难度公式，并把 §7/§8/§8.1
// 里给出的每一组数字做成可核对的量。
//
// 与 Python 版的差异显式落地：
//   - Python 的默认参数用 Go 的显式入参替代（NsCount/Ports/... 一律传入）；
//   - TTL 为 0 表示「退化口径：有效 TTL 取窗口 W」；
//   - 时间一律用 float64 秒。
package main

import "math"

// Params 对应 RFC 5452 §7.1 的符号。
type Params struct {
	NsCount    float64 // N：域的权威服务器数量（约 2.5）
	Ports      float64 // P：解析器可用的源端口数量（常为 1）
	IDSpace    float64 // I：可用的 Query ID 数量（最大 65536）
	Window     float64 // W：机会窗口，秒（常为 0.1）
	Outstanding float64 // D：相同的在途查询数（通常为 1）
}

// DefaultParams 是 RFC 5452 §7.1 列出的常用值。
func DefaultParams() Params {
	return Params{NsCount: 2.5, Ports: 1, IDSpace: 65536, Window: 0.1, Outstanding: 1}
}

// ProblemSpace 返回 N*P*I。
func (p Params) ProblemSpace() float64 { return p.NsCount * p.Ports * p.IDSpace }

// PSingle 是 P_s = D*F/(N*P*I)，F 为窗口内到达的伪造包数。
func (p Params) PSingle(fakePackets float64) float64 {
	return p.Outstanding * fakePackets / p.ProblemSpace()
}

// PSingleFromRate 是 F = R*W 之后的 P_s。
func (p Params) PSingleFromRate(ratePps float64) float64 {
	return p.PSingle(ratePps * p.Window)
}

// ReducedConstant 是 §7.2 化简后的分母 N*P*I/(D*W)，文档给的是 1638400。
func (p Params) ReducedConstant() float64 {
	return p.ProblemSpace() / (p.Outstanding * p.Window)
}

// Attempts 是 A = T/TTL；TTL 为 0 时按退化口径取 W。
func (p Params) Attempts(periodS, ttlS float64) float64 {
	effTTL := ttlS
	if effTTL == 0 {
		effTTL = p.Window
	}
	return periodS / effTTL
}

// PCombined 是 P_cs = 1 - (1 - P_s)^A。
func (p Params) PCombined(periodS, ttlS, ratePps float64) float64 {
	ps := p.PSingleFromRate(ratePps)
	return 1.0 - math.Pow(1.0-ps, p.Attempts(periodS, ttlS))
}

// PeriodForProbability 用精确公式反解达到 target 所需的时间（秒）。
func (p Params) PeriodForProbability(target, ttlS, ratePps float64) float64 {
	ps := p.PSingleFromRate(ratePps)
	effTTL := ttlS
	if effTTL == 0 {
		effTTL = p.Window
	}
	return math.Log(1.0-target) / math.Log(1.0-ps) * effTTL
}

// RateForProbability 用精确公式反解达到 target 所需的发包速率（pps）。
func (p Params) RateForProbability(target, periodS, ttlS float64) float64 {
	a := p.Attempts(periodS, ttlS)
	ps := 1.0 - math.Pow(1.0-target, 1.0/a)
	return ps * p.ProblemSpace() / (p.Outstanding * p.Window)
}

// MeanTriesSingleShot 是 §7 首段的 N*P*I/2。
func (p Params) MeanTriesSingleShot() float64 { return p.ProblemSpace() / 2.0 }

// BirthdayPSingle 是 §5 的线性放大：D 个相同在途查询。
func BirthdayPSingle(basePSingle, outstanding float64) float64 {
	return basePSingle * outstanding
}

const dnsResponseBytes = 80

// BandwidthBps 把 pps 换成 bit/s，文档假设的最小应答是 80 字节。
func BandwidthBps(ratePps float64) float64 {
	return ratePps * dnsResponseBytes * 8.0
}

// PortRandomizationGain 源端口随机化把问题空间放大的倍数。
func PortRandomizationGain(ports float64) float64 { return ports }
